crates/hir-ty/src/infer/expr.rs

```diff
@@ -262,19 +262,44 @@ impl<'a> InferenceContext<'a> {
                 );
 
                 // Now go through the argument patterns
-                for (arg_pat, arg_ty) in args.iter().zip(sig_tys) {
+                for (arg_pat, arg_ty) in args.iter().zip(&sig_tys) {
                     self.infer_pat(*arg_pat, &arg_ty, BindingMode::default());
                 }
 
                 let prev_diverges = mem::replace(&mut self.diverges, Diverges::Maybe);
                 let prev_ret_ty = mem::replace(&mut self.return_ty, ret_ty.clone());
 
-                self.infer_expr_coerce(*body, &Expectation::has_type(ret_ty));
+                let inner_ty = self.infer_expr_coerce(*body, &Expectation::has_type(ret_ty));
+
+                let inner_ty = if *is_async {
+                    // Use the first type parameter as the output type of future.
+                    // existential type AsyncBlockImplTrait<InnerType>: Future<Output = InnerType>
+                    let impl_trait_id =
+                        crate::ImplTraitId::AsyncBlockTypeImplTrait(self.owner, *body);
+                    let opaque_ty_id = self.db.intern_impl_trait_id(impl_trait_id).into();
+                    TyKind::OpaqueType(opaque_ty_id, Substitution::from1(Interner, inner_ty))
+                        .intern(Interner)
+                } else {
+                    inner_ty
+                };
 
                 self.diverges = prev_diverges;
                 self.return_ty = prev_ret_ty;
 
-                closure_ty
+                sig_tys.pop();
+                sig_tys.push(inner_ty);
+
+                let sig_ty = TyKind::Function(FnPointer {
+                    num_binders: 0,
+                    sig: FnSig { abi: (), safety: chalk_ir::Safety::Safe, variadic: false },
+                    substitution: FnSubst(
+                        Substitution::from_iter(Interner, sig_tys.clone()).shifted_in(Interner),
+                    ),
+                })
+                .intern(Interner);
+
+                TyKind::Closure(closure_id, Substitution::from1(Interner, sig_ty.clone()))
+                    .intern(Interner)
```

Comment:
I don't know if the previous result of `infer_expr_coerce` was unused for a reason, but this PR also replaces `ret_ty` with `inner_ty` (returned from `infer_expr_coerce`) and remakes `sig_ty` and `closure_ty`, even for non-async closures.

Reply:
I believe as `ret_ty` could be a type variable, the unification in `infer_expr_coerce` would change the closures return type as well.
So this reconstruction should only be necessary for the async case